AST/Sema: Make `MemberImportVisibility` a migratable feature

tshortli opened the pull request:
The migration to `MemberImportVisibility` can be performed mechanically by adding missing import declarations, so offer automatic migration for the feature.

Also fixes a bug in the migratable feature infrastructure where `hasFeature()` would return `false` incorrectly when a feature is enabled for migration.

Resolves rdar://151931597.

tshortli committed:
Basic: Fix a bug in LangOptions::hasFeature().

Don't skip checking if a feature is enabled for migration when the feature also
has an associated language version.

tshortli committed:
AST/Sema: Make MemberImportVisibility a migratable feature.

The migration to `MemberImportVisibility` can be performed mechanically by
adding missing import declarations, so offer automatic migration for the
feature.

Resolves rdar://151931597.

tshortli commented:
@swift-ci please smoke test

tshortli commented:
@xedin @AnthonyLatsis @DougGregor I'd appreciate a review as I plan to cherry-pick this to `release/6.2`.

tshortli commented:
@swift-ci please smoke test

tshortli commented:
@swift-ci please smoke test

xedin commented on lib/Sema/TypeCheckNameLookup.cpp:
```diff
@@ -920,22 +920,31 @@ class MissingImportFixItCache {
 
+  bool isMigrating =
+      ctx.LangOpts.getFeatureState(Feature::MemberImportVisibility)
+          .isEnabledForMigration();
```
@AnthonyLatsis is it currently possible to coalesce the fix-it from multiple different diagnostics (possibilities here are - if I understand correctly - the same id but different arguments, locations, different ids), or we'd just end up adding multiple `import <<module>>` from that?

xedin replied:
Yes, something like that, let's wait for @AnthonyLatsis to confirm but I have a suspicion that we might not be able to coalesce the fix-its...

tshortli replied:
I've implemented the suggested approach where in migration mode one warning is emitted per module that needs to be imported. Even if the migration infrastructure can handle the duplicate diagnostics that were emitted previously, I think this is going to yield the best developer experience.

xedin replied:
LGTM, thank you!

tshortli commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -167,10 +167,10 @@ ERROR(init_candidate_inaccessible,none,
       "%kind0 is not available due to missing import of defining module %1",
       (const ValueDecl *, const ModuleDecl *))
-ERROR(candidate_from_missing_imports_2_or_more,none,
+GROUPED_ERROR(member_from_missing_imports_2_or_more,MemberImportVisibility,none,
```
Yeah, I downgrade it explicitly in migration mode.

AnthonyLatsis commented on lib/Sema/TypeCheckNameLookup.cpp:
```diff
@@ -1039,8 +1048,7 @@ bool swift::maybeDiagnoseMissingImportForMember(const ValueDecl *decl,
     if (modulesToImport.empty())
       return false;
 
-    diagnoseMissingImportsForMember(decl, modulesToImport, sf, loc);
-    return true;
+    return diagnoseMissingImportsForMember(decl, modulesToImport, sf, loc);
```
The doc comment for `maybeDiagnoseMissingImportForMember` needs an update for when the function returns true now that we return false on warnings. I think it would also help to copy-paste that part onto `diagnoseMissingImportsForMember`.

AnthonyLatsis reviewed (approved):
Looks great! Thanks, Allan.

tshortli committed:
AST/Sema: Remove code adding `@_spi` to suggested import fix-its.

The was never invoked because inaccessibility due to SPI protection level is
always diagnosed before missing imports are diagnosed. The functionality could
therefore not be tested and should be removed.

tshortli committed:
AST: Simplify the interface of `DiagnosticEngine::getBestAddImportFixItLoc()`.

For clarity, it should just take a `SourceFile`.

tshortli commented:
There have been significant updates, so please take a look again if you reviewed this previously.

DougGregor reviewed (approved):
This looks fantastic now, thank you

zhrispineda commented on lib/Sema/TypeCheckNameLookup.cpp:
```diff
@@ -1047,7 +1044,76 @@ bool swift::maybeDiagnoseMissingImportForMember(const ValueDecl *decl,
+    }
+  }
+
+  // Emit one warning for each module that needcs to be imported and emit notes
```
Typo: needcs
